components/eam/src/chemistry/modal_aero/seasalt_model.F90

```diff
@@ -475,68 +503,67 @@ subroutine seasalt_emis(u10, u10cubed, lchnk, srf_temp, ocnfrc, ncol, cflx, emis
     !                 of mode with mass and number in MOM modes
     ! Mixing state 1: external mixture, add OM to mass and number
 
-   emit_this_mode = (/ .true., .true., .false. /)
 
    ! Loop over OM modes
-   om_num_mode_loop: do m_om=1,om_num_modes ! modes in which to emit OM
+   om_num_mode_loop: do ispec = 1, om_num_modes ! modes in which to emit OM
       if (size(om_num_ind) .eq. 1) then
          call endrun(&
          "Error: om_num_ind is a scalar, but attempting to calculate MOM.  Something bad happened!!  We should never get here!")
       endif
       
-      m = om_num_ind(m_om)
-      mn=seasalt_indices(nslt+nslt_om+m)
+      m = om_num_ind(ispec)
+      num_mode_idx=seasalt_indices(nslt+nslt_om+m)
 
       ! add number tracers for organics-only modes
-      if (emit_this_mode(m_om)) then
-         section_loop_OM_num: do i=1, nsections
-            cflx_help2(:ncol) = 0.0_r8
-            if (Dg(i) >= sst_sz_range_lo(nslt+m_om) .and. Dg(i) < sst_sz_range_hi(nslt+m_om)) then
-               cflx_help2(:ncol)=fi(:ncol,i)*ocnfrc(:ncol)*emis_scale
+      if (emit_this_mode(ispec)) then
+         section_loop_OM_num: do ibin=1, nsections
+            cflx_tmp1(:ncol) = 0.0_r8
+            if (Dg(ibin) >= sst_sz_range_lo(nslt+ispec) .and. Dg(ibin) < sst_sz_range_hi(nslt+ispec)) then
+               cflx_tmp1(:ncol)=fi(:ncol,ibin)*ocnfrc(:ncol)*emis_scale
 
                ! Mixing state 3: internal mixture, add OM to mass and number
-               cflx(:ncol,mn) = cflx(:ncol,mn) + cflx_help2(:ncol) * &
-                               (1._r8 / (1._r8 - om_ssa(:ncol, i)) - 1._r8)
+               cflx(:ncol,num_mode_idx) = cflx(:ncol,num_mode_idx) + cflx_tmp1(:ncol) * &
+                                          (1._r8 / (1._r8 - om_ssa(:ncol, ibin)) - 1._r8)
             endif
          enddo section_loop_OM_num
       endif
    enddo om_num_mode_loop
 
-   om_mode_loop: do m_om=1,nslt_om
-      mm = seasalt_indices(nslt+m_om)
+   om_mode_loop: do ispec=1,nslt_om
+      mass_mode_idx = seasalt_indices(nslt+ispec)
 
-      cflx(:ncol,mm)=0.0_r8
+      cflx(:ncol,mass_mode_idx)=0.0_r8
       
-      if (emit_this_mode(m_om)) then
+      if (emit_this_mode(ispec)) then
          ! write(iulog,"(A30,A10,I3)") "Constituent name and number: ", trim(seasalt_names(nslt+m_om)), mm ! for debugging
          ! add mass tracers
-         om_type_loop: do n=1,n_org
-            section_loop_OM_mass: do i=1, nsections
-               if (Dg(i)>=sst_sz_range_lo(nslt+m_om) .and. Dg(i)<sst_sz_range_hi(nslt+m_om)) then
-                  cflx_help2(:ncol)=fi(:ncol,i)*ocnfrc(:ncol)*emis_scale &
-                                    *4._r8/3._r8*pi*rdry(i)**3*dns_aer_sst  ! should use dry size, convert from number to mass flux (kg/m2/s)
+         om_type_loop: do iorg = 1, n_org
+            section_loop_OM_mass: do ibin = 1, nsections
+               if (Dg(ibin)>=sst_sz_range_lo(nslt+ispec) .and. Dg(ibin)<sst_sz_range_hi(nslt+ispec)) then
```

Comment:
This looks similar to that in seasalt_emis. is it possible to make it a subroutine that can be called in both seasalt and MOM emission?

Reply:
further refactor seasalt_emis, added seasalt_emisflx_calc for number and mass fluxes calculations.